packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -0,0 +1,460 @@
+
+    return new CwdFS(unplugPath);
+  }
+
+  private isUnplugged(ident: Ident, dependencyMeta: DependencyMeta) {
+    if (dependencyMeta.unplugged)
+      return true;
+
+    if (FORCED_UNPLUG_PACKAGES.has(ident.identHash))
+      return true;
+
+    return false;
+  }
+}
+
+const persistDirEntry = (parentPath: PortablePath, entryName: string, baseFs: PortableNodeModulesFS, mtime: Date, children: NodeModulesTimestampsTree, dirent?: Dirent) => {
+  const fullPath = ppath.join(parentPath, toFilename(entryName));
+  if (!dirent || dirent.isDirectory()) {
+    xfs.mkdirSync(fullPath);
+
+    const entries = baseFs.readdirSync(fullPath, {withFileTypes: true});
+    for (const entry of entries) {
+      const node = children.get(toFilename(entry.name));
+      const entryMtime = node ? node.mtime : mtime;
+      const entryChildren = node ? node.children : children;
+      persistDirEntry(fullPath, entry.name, baseFs, entryMtime, entryChildren, entry);
+    }
+
+    xfs.utimesSync(fullPath, mtime, mtime);
+    xfs.chmodSync(fullPath, 0o777);
+  } else if (dirent) {
+    if (dirent.isFile()) {
+      const content = baseFs.readFileSync(fullPath);
+      xfs.writeFileSync(fullPath, content);
+    } else if (dirent.isSymbolicLink()) {
+      const target = baseFs.readlinkSync(fullPath);
+      xfs.symlinkSync(target, fullPath);
+    } else {
+      throw new Error(`Unsupported file type (file: ${fullPath})`);
+    }
+  }
+};
+
+const wasModuleDirChanged = (dir: PortablePath, stats: Stats, mtime: Date, ignoreRegexp: RegExp, excludeNodeModules: boolean): boolean => {
+  if (+stats.mtime !== +mtime) {
+    return true;
+  } else {
+    const entries = xfs.readdirSync(dir);
+    for (const entry of entries) {
+      const entryPath = ppath.join(dir, entry);
+      const stats = xfs.lstatSync(entryPath);
+      if (stats.isDirectory() && !ignoreRegexp.test(entryPath) && (!excludeNodeModules || entry !== 'node_modules') && wasModuleDirChanged(entryPath, stats, mtime, ignoreRegexp, false)) {
+        return true;
+      }
+    }
+    return false;
+  }
+};
+
+const persistNodeModules = (rootPath: PortablePath, nodeModulesFS: PortableNodeModulesFS): NodeModulesTimestampsTree => {
+  const newTsTree = new Map();
+  const tsTree = nodeModulesFS.getNodeModulesDirTimestamps();
+  const NODE_MODULES = toFilename('node_modules');
+  const DEL_BLACKLIST = /\/(node_modules\/\.|__ivy_ngcc__(\/|$))/;
+
+  const persistNode = (parentPath: PortablePath, dir: Filename, mtime: Date, children: NodeModulesTimestampsTree) => {
+    const nodePath = ppath.join(parentPath, dir);
+    let stats;
+    try {
+      stats = xfs.lstatSync(nodePath);
+    } catch (e) {
+      if (e.code !== 'ENOENT') {
+        throw e;
+      }
+    }
+    // Non existing directory
+    if (!stats) {
+      persistDirEntry(parentPath, dir, nodeModulesFS, mtime, children);
+      // Existing directory
+    } else {
+      const hasInnerNodeModules = (children.size === 1 && children.has(NODE_MODULES));
```

Comment:
I think it would be useful to document in which practical case will each of the code branches be reached. It's hard to understand what special cases are covered 🤔

In particular here, I'm not sure I understand why the number of children matters...